### PR TITLE
[feat]: [CDS-58214]: replace environmentRef field name to envIdentifer for serviceOverrideNG

### PR DESCRIPTION
**Pull request by tchaurasiya**
## Describe your changes
1. Migration to in-place change environtmentRef field name to envIdentifier
2. Added new envIdentifier in NGServiceOverridesEntity.java, updated index in java file as well in index.txt (Keeping the old field to be safegurad for migration failure)
3. Removing javax NotNull from environtmentRef & envIdentifier fields in NGServiceOverridesEntity
4. Not changing any internal POJO class yet, Only change in NGServiceOverridesEntity
5. Any upsert query will only update envIdentifier
6. Every Criteria for get or update will contain orOperation of Criteria.where(environmentRef), Criteria.where(envIdentifier)

**Sanity test plan**
Project, org, Account level cruds
Delete all on project delete
Older entities with envRef exist in DB
New entities with envId exist in DB
Mixed entities with envId, environmentRef exist in DB
ServiceOverride View page
One pipeline execution involving overrides

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>

You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger utAll`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/45893)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/migration/background/ChangeEnvironmentRefFieldNameToEnvIdentifierMigration.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+                                            .map(AccountDTO::getIdentifier)
+                                            .collect(Collectors.toList());
+
+      accountIdentifiers.forEach(accountId -> {
+        try {
+          log.info(DEBUG_LOG
+              + "Starting migration of changing environmentRef field name to envIdentifier in serviceOverrideNG for account : "
+              + accountId);
+
+          Update update = new Update().rename(OLD_FIELD_NAME, NEW_FIELD_NAME);
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/migration/background/ChangeEnvironmentRefFieldNameToEnvIdentifierMigration.java:43]: will this retain the older fields or rename in-place ?

Comment (yogesh-chauhan) [reply 120-ng-manager/src/main/java/io/harness/ng/core/migration/background/ChangeEnvironmentRefFieldNameToEnvIdentifierMigration.java:43]: The
$rename
operator logically performs an $unset of both the old name and the new name, and then performs a $set operation with the new name

**Comment by tchaurasiya** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/migration/background/ChangeEnvironmentRefFieldNameToEnvIdentifierMigration.java` line 43 at r1](https://reviewable.io/reviews/harness/harness-core/45893#-NRmJNUccl2qeCHAnZik-r1-43:-NRqhr4P9-fyT2lclXL4:b9rwhu8) ([raw file](https://github.com/harness/harness-core/blob/7c0da64460e991f87f6eea567a689ae4bf99155e/120-ng-manager/src/main/java/io/harness/ng/core/migration/background/ChangeEnvironmentRefFieldNameToEnvIdentifierMigration.java#L43)):*
<details><summary><i>Previously, yogesh-chauhan (Yogesh Chauhan) wrote…</i></summary><blockquote>

The
$rename
operator logically performs an $unset of both the old name and the new name, and then performs a $set operation with the new name
</blockquote></details>

Yes. Eventual result is in-place change.
But internally doesn't seem like it do unset and set.



_Code snippet:_
```Java
The $rename operator in MongoDB works by modifying the document's internal BSON structure to change the name of the specified field.
```

**Comment by tchaurasiya** [comment]
trigger messagecheck

**Comment by yogesh-chauhan** [comment]
*[`120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java` line 785 at r3](https://reviewable.io/reviews/harness/harness-core/45893#-NS5Xyaa9K6N1ewjwNGT:-NS5Xyaa9K6N1ewjwNGU:b42pry5) ([raw file](https://github.com/harness/harness-core/blob/868cb84e99db2de7006670de2a02acf0ccc6960c/120-ng-manager/src/main/java/io/harness/ng/core/environment/resources/EnvironmentResourceV2.java#L785)):*
> ```Java
>                 serviceOverridesEntity.getProjectIdentifier(),
>                 isNotBlank(serviceOverridesEntity.getEnvIdentifier()) ? serviceOverridesEntity.getEnvIdentifier()
>                                                                       : serviceOverridesEntity.getEnvironmentRef(),
> ```

we can push this method to the serviceOverridesEntity itself like fetchEnvironmentRef..

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 3 files at r1, 1 of 9 files at r2.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/45893)* status: Some of the checks are too old. 13 file(s) to review. 1. unresolved discussions. 0 of 2 approvals obtained.

**Pull request closed without merging**